include/k4arecord/record.hpp

```diff
@@ -0,0 +1,280 @@
+    {
+        k4a_result_t result = k4a_record_add_custom_subtitle_track(m_handle,
+                                                                   track_name,
+                                                                   codec_id,
+                                                                   codec_context,
+                                                                   codec_context_size,
+                                                                   track_settings);
+
+        if (K4A_RESULT_SUCCEEDED != result)
+        {
+            throw error("Failed to add custom subtitle track!");
+        }
+    }
+
+    /** Writes the recording header and metadata to file
+     * Throws error on failure
+     *
+     * \sa k4a_record_write_header
+     */
+    void write_header()
+    {
+        k4a_result_t result = k4a_record_write_header(m_handle);
+
+        if (K4A_RESULT_SUCCEEDED != result)
+        {
+            throw error("Failed to write header!");
+        }
+    }
+
+    /** Writes a camera capture to file
+     * Throws error on failure
+     *
+     * \sa k4a_record_write_capture
+     */
+    void write_capture(capture &capture)
+    {
+        k4a_result_t result = k4a_record_write_capture(m_handle, capture.handle());
+
+        if (K4A_RESULT_SUCCEEDED != result)
+        {
+            throw error("Failed to write capture!");
+        }
+    }
+
+    /** Writes an imu sample to file
+     * Throws error on failure
+     *
+     * \sa k4a_record_write_imu_sample
+     */
+    void write_imu_sample(k4a_imu_sample_t &imu_sample)
+    {
+        k4a_result_t result = k4a_record_write_imu_sample(m_handle, imu_sample);
+
+        if (K4A_RESULT_SUCCEEDED != result)
+        {
+            throw error("Failed to write imu sample!");
+        }
+    }
+
+    /** Writes data for a custom track to file
+     * Throws error on failure
+     *
+     * \sa k4a_record_write_custom_track_data
+     */
+    void write_custom_track_data(const char *track_name,
+                                 uint64_t device_timestamp_usec,
+                                 uint8_t *custom_data,
+                                 size_t custom_data_size)
+    {
+        k4a_result_t result = k4a_record_write_custom_track_data(m_handle,
+                                                                 track_name,
+                                                                 device_timestamp_usec,
+                                                                 custom_data,
+                                                                 custom_data_size);
+
+        if (K4A_RESULT_SUCCEEDED != result)
+        {
+            throw error("Failed to write imu sample!");
+        }
+    }
+
+    /** Opens a new recording file for writing
+     * Throws error on failure
+     *
+     * \sa k4a_record_create
+     */
+    static record create(const char *path, device &device, k4a_device_configuration_t &device_configuration)
```

Comment:
I wondered about why k4a::device don't hold k4a_device_configuration_t?
If it is holded, this function doesn't need to input k4a_device_configuration_t.
(Or, I think it is good add function to get k4a_device_configuration_t from k4a_device_t.)

Reply:
Storing the configuration would allow the create API for record to eliminate one parameter, but the SDK would have to keep track if k4a_device_start_cameras has been called to provide the config. If the user calls record::create before start camera's then should it it just fail.

Another issue we would have to be concerned with is after ::stop_cameras() is called and the configuration is known by the SDK. Simple enough to null out, but we would need to test and ensure that works as expected. 

I think for now it would simplier to just pass in the config and not create any dependencies between the two API's and just pass the parameter into each API.

Reply:
I agree with Wes on this. The current API also allows for recordings to be generated without a device present, which is useful for synthetic data testing.

Reply:
I got it. Thanks,